File: src/test/java/com/pubnub/api/endpoints/DeleteMessagesEndpointTest.java
package com.pubnub.api.endpoints;

import com.github.tomakehurst.wiremock.junit.WireMockRule;
import com.pubnub.api.PubNub;
import com.pubnub.api.PubNubException;
import com.pubnub.api.models.consumer.history.PNDeleteMessagesResult;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import java.io.IOException;
import java.util.Arrays;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.options;
import static org.junit.Assert.assertNotNull;


public class DeleteMessagesEndpointTest extends TestHarness {

    @Rule
    public WireMockRule wireMockRule = new WireMockRule(options().port(this.PORT), false);

    private DeleteMessages partialHistory;
    private PubNub pubnub;

    @Before
    public void beforeEach() throws IOException, PubNubException {
        pubnub = this.createPubNubInstance();
        partialHistory = pubnub.deleteMessages();
        wireMockRule.start();
    }

    @After
    public void afterEach() {
        pubnub.destroy();
        pubnub = null;
        wireMockRule.stop();
    }

    @Test
    public void testSyncSuccess() throws PubNubException {
        stubFor(delete(urlPathEqualTo("/v3/history/sub-key/mySubscribeKey/channel/mychannel,my_channel"))
                .willReturn(aResponse().withBody("{\"status\": 200, \"error\": False, \"error_message\": \"\"}")));

        PNDeleteMessagesResult response = partialHistory.channels(Arrays.asList("mychannel,my_channel")).sync();

        assertNotNull(response);
    }

    @Test
    public void testSyncAuthSuccess() throws PubNubException {
        stubFor(delete(urlPathEqualTo("/v3/history/sub-key/mySubscribeKey/channel/mychannel,my_channel"))
                .willReturn(aResponse().withBody("{\"status\": 200, \"error\": False, \"error_message\": \"\"}")));

        pubnub.getConfiguration().setAuthKey("authKey");

        PNDeleteMessagesResult response = partialHistory.channels(Arrays.asList("mychannel,my_channel")).sync();

        assertNotNull(response);
    }

    @Test
    public void testFailure() throws PubNubException {
        stubFor(delete(urlPathEqualTo("/v3/history/sub-key/mySubscribeKey/channel/mychannel,my_channel"))
                .willReturn(aResponse().withBody("{\"status\": 403, \"error\": False, \"error_message\": \"wut\"}")));

        pubnub.getConfiguration().setAuthKey("authKey");

        try {
            partialHistory.channels(Arrays.asList("mychannel,my_channel")).sync();
        } catch (PubNubException ex) {
            assert (ex.getErrormsg().equals("wut"));
        }
    }

}

File: src/test/kotlin/com/pubnub/api/legacy/endpoints/DeleteMessagesEndpointTest.kt
package com.pubnub.api.legacy.endpoints

import com.github.tomakehurst.wiremock.client.WireMock.aResponse
import com.github.tomakehurst.wiremock.client.WireMock.delete
import com.github.tomakehurst.wiremock.client.WireMock.deleteRequestedFor
import com.github.tomakehurst.wiremock.client.WireMock.findAll
import com.github.tomakehurst.wiremock.client.WireMock.stubFor
import com.github.tomakehurst.wiremock.client.WireMock.urlMatching
import com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo
import com.pubnub.api.CommonUtils.failTest
import com.pubnub.api.enums.PNOperationType
import com.pubnub.api.legacy.BaseTest
import org.awaitility.Awaitility
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class DeleteMessagesEndpointTest : BaseTest() {

    @Test
    fun testSyncSuccess() {
        stubFor(
            delete(urlPathEqualTo("/v3/history/sub-key/mySubscribeKey/channel/mychannel,my_channel"))
                .willReturn(
                    aResponse().withBody(
                        """
                        {
                          "status": 200,
                          "error": false,
                          "error_message": ""
                        }
                        """.trimIndent()
                    )
                )
        )

        pubnub.deleteMessages(
            channels = listOf("mychannel,my_channel")
        ).sync()!!
    }

    @Test
    fun testSyncAuthSuccess() {
        stubFor(
            delete(urlPathEqualTo("/v3/history/sub-key/mySubscribeKey/channel/mychannel,my_channel"))
                .willReturn(
                    aResponse().withBody(
                        """
                        {
                          "status": 200,
                          "error": false,
                          "error_message": ""
                        }
                        """.trimIndent()
                    )
                )
        )

        pubnub.configuration.authKey = "authKey"

        pubnub.deleteMessages(
            channels = listOf("mychannel,my_channel")
        ).sync()!!

        val requests = findAll(deleteRequestedFor(urlMatching("/.*")))
        assertEquals(1, requests.size)
        assertEquals("authKey", requests[0].queryParameter("auth").firstValue())
    }

    @Test
    fun testFailure() {
        stubFor(
            delete(urlPathEqualTo("/v3/history/sub-key/mySubscribeKey/channel/mychannel,my_channel"))
                .willReturn(
                    aResponse().withBody(
                        """
                        {
                          "status": 403,
                          "error": false,
                          "error_message": "wut"
                        }
                        """.trimIndent()
                    )
                )
        )

        try {
            pubnub.deleteMessages(
                channels = listOf("mychannel,my_channel")
            ).sync()!!
        } catch (e: Exception) {
            failTest()
        }
    }

    @Test
    fun testAsyncSuccess() {
        stubFor(
            delete(urlPathEqualTo("/v3/history/sub-key/mySubscribeKey/channel/mychannel,my_channel"))
                .willReturn(
                    aResponse().withBody(
                        """
                        {
                          "status": 200,
                          "error": false,
                          "error_message": ""
                        }
                        """.trimIndent()
                    )
                )
        )

        val success = AtomicBoolean()

        pubnub.deleteMessages(
            channels = listOf("mychannel,my_channel")
        ).async { result, status ->
            result!!
            assertFalse(status.error)
            assertEquals(PNOperationType.PNDeleteMessagesOperation, status.operation)
            success.set(true)
        }

        Awaitility.await().atMost(5, TimeUnit.SECONDS).untilTrue(success)
    }
}
